util/env_windows.cc

```diff
@@ -0,0 +1,920 @@
+// Copyright (c) 2017 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+#include <io.h>
+#include <stdio.h>
+#include <wchar.h>
+#include <atomic>
+#include <chrono>
+#include <deque>
+#include <functional>
+#include <memory>
+#include <set>
+#include <thread>
+#include "leveldb/env.h"
+#include "port/port.h"
+#include "util/mutexlock.h"
+#include "util/file_logger.h"
+#include "util/env_windows_test_helper.h"
+
+// Win32 defines the DeleteFile macro
+// Include last, and immediately undefine macro.
+#include <windows.h>
+#ifdef DeleteFile
+#undef DeleteFile
+#endif
+
+namespace leveldb {
+
+namespace {
+
+static int open_read_only_file_limit = -1;
+static int mmap_limit = -1;
+static port::Mutex pread_mutex;
+
+static Status IOError(const std::string& context, int err_number) {
+  return Status::IOError(context, strerror(err_number));
+}
+
+static Status WindowsError(const std::string & context, DWORD err_number)
+{
+  std::string err_mess;
+  LPSTR text = NULL;
+  if (!::FormatMessageA(
+    FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_ALLOCATE_BUFFER,
+    0,
+    err_number,
+    0,
+    (LPTSTR)&text,
+    32000,
+    0
+  )) {
+    err_mess = "unknown error";
+  } else {
+    err_mess = text;
+    ::LocalFree(text);
+  }
+
+  // both error codes map to ENOENT
+  if (err_number == ERROR_FILE_NOT_FOUND ||
+      err_number == ERROR_PATH_NOT_FOUND) {
+    return Status::NotFound(context, err_mess);
+  } else {
+    return Status::IOError(context, err_mess);
+  }
+}
+
+
+std::vector<char> Utf16ToUtf8(const std::wstring& wide)
+{
+  if (wide.empty()) {
+    return std::vector<char>();
+  }
+  int size = WideCharToMultiByte(
+    CP_UTF8,
+    0,
+    &wide[0],
+    wide.size(),
+    NULL,
+    0,
+    NULL,
+    NULL
+  );
+  std::vector<char> buffer(size+1);
+  buffer[size] = '\0';
+  WideCharToMultiByte(CP_ACP, 0, &wide[0], wide.size(), &buffer[0], size, NULL, NULL);
+  return buffer;
+  }
+
+
+std::vector<wchar_t> Utf8ToUtf16(const std::string& narrow)
+{
+  if (narrow.empty()) {
+    return std::vector<wchar_t>();
+  }
+  int size = MultiByteToWideChar(
+    CP_ACP,
+    0,
+    &narrow[0],
+    narrow.size(),
+    NULL,
+    0
+  );
+  std::vector<wchar_t> buffer(size+1);
+  buffer[size] = L'\0';
+  MultiByteToWideChar(CP_UTF8, 0, &narrow[0], narrow.size(), &buffer[0], size);
+  return buffer;
+}
+
+// Helper class to limit resource usage to avoid exhaustion.
+// Currently used to limit read-only file descriptors and mmap file usage
+// so that we do not end up running out of file descriptors, virtual memory,
+// or running into kernel performance problems for very large databases.
+class Limiter {
```

Comment:
We could consider moving Limiter into a helper module of its own (util/limiter.{h,cc}).  Are there any differences from the Posix version?